core/src/main/java/org/apache/iceberg/util/TableScanUtil.java

```diff
@@ -56,7 +57,10 @@ public static boolean hasDeletes(FileScanTask task) {
     Preconditions.checkArgument(lookback > 0, "Invalid split planning lookback (negative or 0): %s", lookback);
     Preconditions.checkArgument(openFileCost >= 0, "Invalid file open cost (negative): %s", openFileCost);
 
-    Function<FileScanTask, Long> weightFunc = file -> Math.max(file.length(), openFileCost);
+    // Check the size of delete file as well to avoid unbalanced bin-packing
+    Function<FileScanTask, Long> weightFunc = file -> Math.max(
+        file.length() + file.deletes().stream().mapToLong(ContentFile::fileSizeInBytes).sum(),
```

Comment:
Should also consider the cost when using different join algorithm between delete files and data files ? 

For the equality files, the cost to join data file is :  `file.recordCount() * sum(eqDeleteFile.recordCount()) *  avgRecordByteSize)` .

For the positional delete files, the cost to join data file is: `file.length() + sum(posFiles.fileSizeInBytes())`.

The current approach sounds like we are treating all the delete files are positional delete files...

Reply:
Thanks for your reply @openinx . Actually I don't think it's necessary to separate delete file to eq-delete and pos-delete when calculating cost. besides the algorithm here would be complicated, the top-level logic to apply delete files is 
`applyEqDeletes(applyPosDeletes(records))`
Generally I think the base logic above for eq-delete and pos-delete is the same.